src/bin/pg_autoctl/pgsql.c

```diff
@@ -270,65 +519,107 @@ pgsql_retry_open_connection(PGSQL *pgsql)
 				if (PQstatus(connection) == CONNECTION_OK)
 				{
 					connectionOk = true;
-					log_info("Successfully connected to \"%s\"",
-							 pgsql->connectionString);
+					log_info("Successfully connected to \"%s\" "
+							 "after %d attempts in %d seconds.",
+							 pgsql->connectionString,
+							 pgsql->retryPolicy.attempts,
+							 (int) (now - startTime));
 				}
 				else
 				{
-					log_error("Failed to connect after successful "
-							  "ping, please verify authentication "
-							  "and logs on the server at \"%s\"",
-							  pgsql->connectionString);
+					lastWarningMessage = PQPING_OK;
+					lastWarningTime = now;
+
+					(void) log_connection_error(connection, LOG_WARN);
+
+					log_warn("Failed to connect after successful "
+							 "ping, please verify authentication "
+							 "and logs on the server at \"%s\"",
+							 pgsql->connectionString);
+
+					log_warn("Authentication might have failed on the Postgres "
+							 "server due to missing HBA rules.");
 				}
 				break;
 			}
 
+			/*
+			 * https://www.postgresql.org/docs/current/libpq-connect.html
+			 *
+			 * The server is running but is in a state that disallows
+			 * connections (startup, shutdown, or crash recovery).
+			 */
 			case PQPING_REJECT:
 			{
-				log_error("Connection rejected: \"%s\"",
-						  pgsql->connectionString);
-				retry = false;
+				if (lastWarningMessage != PQPING_REJECT ||
+					(now - lastWarningTime) > 30)
+				{
+					lastWarningMessage = PQPING_REJECT;
+					lastWarningTime = now;
+
+					log_warn(
+						"The server at \"%s\" is running but is in a state "
+						"that disallows connections (startup, shutdown, or "
+						"crash recovery).",
+						pgsql->connectionString);
+				}
 				break;
 			}
 
+			/*
+			 * https://www.postgresql.org/docs/current/libpq-connect.html
+			 *
+			 * The server could not be contacted. This might indicate that the
+			 * server is not running, or that there is something wrong with the
+			 * given connection parameters (for example, wrong port number), or
+			 * that there is a network connectivity problem (for example, a
+			 * firewall blocking the connection request).
+			 */
 			case PQPING_NO_RESPONSE:
 			{
-				int logLevel = attempts % 30 == 0 ? LOG_INFO : LOG_DEBUG;
-
-				log_level(logLevel,
-						  "PQping: no response from server at \"%s\" "
-						  "after %d attempts",
-						  pgsql->connectionString,
-						  attempts);
-				retry = true;
+				if (lastWarningMessage != PQPING_NO_RESPONSE ||
```

Comment:
I was wondering when we would get this response. Turns out it's because of the `connect_timeout` connection string param or the `PGCONNECT_TIMEOUT` env var. We don't set either of these. On my machine this means the FSM is blocked for 30 seconds when I drop all traffic to the monitor using:
```bash
sudo iptables -A INPUT -p tcp --dport 5432 -m conntrack --ctstate NEW,ESTABLISHED -j DROP
```
This seems like a serious problem as this will impact the 30s timeout semantics of `DEMOTE_TIMEOUT`

Reply:
In other places of the code, like for calling `pg_basebackup`, we force the environment variable to 1 second. I have now done the same for establishing all our connections, always, so that we have better control.